Troubleshooting unbalanced cluster docs (#105774)
elastic/elasticsearch: docs/reference/troubleshooting/troubleshooting-unbalanced-cluster.asciidoc

* shard count
* disk usage
* ingest load forecast for write indices

Please note, elasticsearch does not take into account amount or complexity of search queries when rebalancing shards.

There is no guarantee that individual components are going to be evenly spread across the nodes.
This might appear as some of the nodes have less shards allocated or using less disk
as long as they are assigned with the shards with greater ingest load forecast.

You might get a detailed output of assigned workloads per node using <<cat-allocation,cat allocation command>>.

[source,console,id=cat-allocation-example]
--------------------------------------------------
GET /_cat/allocation
--------------------------------------------------

The API returns the following response:

[source,text]
--------------------------------------------------
shards shards.undesired write_load.forecast disk.indices.forecast disk.indices disk.used disk.avail disk.total disk.percent host      ip        node    node.role
     1                0                 0.0                  260b         260b    47.3gb     43.4gb    100.7gb           46 127.0.0.1 127.0.0.1 CSUXak2 himrst

> --------------------------------------------------

Reviewer: Could we turn this into a docs test, i.e. assert that the structure of the output of this API is as expected?

[[troubleshooting-unbalanced-cluster]]
== Troubleshooting an unbalanced cluster

> Since 8.6 Elasticsearch is balancing shards across data tier to achieve a good compromise between

Reviewer: I'd avoid specifying a version here, it's not very future-proof. The docs are specific to a (minor) version, we can just describe things as they are today.

Reviewer: I don't think the fact that this behaviour changed is relevant to the troubleshooting process, indeed I worry that it could cause more confusion than it solves to mention it. Let's just stick with documenting the present version, it's always ok to say that newer versions have better troubleshooting support.

Reviewer 2: I actually recommended (I think?) mentioning 8.6 as a boundary point for behavior. There was a really big change in 8.6 for balancing decisions, which causes a lot of shard movement on upgrade, so it would be good to reassure users that what they are seeing is expected and not a cause for alarm in and of itself.

Author: > which causes a lot of shard movement on upgrade

I do not think this is accurate. The balanced or nearly balanced clusters are not going to have any shard movements when changing/upgrading allocator. The unbalanced ones likely did had a significant amount of shard movement before upgrade, though it was harder to notice that since there were no way to estimate the remaining moves.

Reviewer: How about mentioning this in the 8.6 release notes?

Reviewer: > I do not think this is accurate. The balanced or nearly balanced clusters are not going to have any shard movements when changing/upgrading allocator.

I think Dianna is right here, the weight function changed in 8.6 to include more variables than just shard count, so a cluster which is balanced according to the <8.6 balancer might indeed not be balanced when upgraded to ≥8.6.

Author: True, however it does not necessarily translates to a lot of movements. I think it is more accurate to represent it as it _might_ cause movements and the _count_ depends on many things.

* shard count
* disk usage
* ingest load forecast for write indices

Please note, elasticsearch does not take into account amount or complexity of search queries when rebalancing shards.

There is no guarantee that individual components are going to be evenly spread across the nodes.
This might appear as some of the nodes have less shards allocated or using less disk

> as long as they are assigned with the shards with greater ingest load forecast.

Reviewer: In terms of troubleshooting, I'd recommend making this more actionable. For instance, this could be step 1 of the troubleshooting process: determine whether your cluster is really unbalanced or not (from Elasticsearch's point of view). If Elasticsearch thinks it's balanced then typically there's no action needed, it will keep all nodes within the disk watermarks (link to the disk allocation docs); if the default disk watermarks are unsatisfactory then change them.

Conversely if ES thinks it's unbalanced then it will be working to balance it, but deliberately quite slowly to avoid disrupting production traffic (unless there is an urgent need, in which case it will speed up). If it struggles to reach a balanced state then try resetting the balance and/or increasing the threshold.

> You might get a detailed output of assigned workloads per node using <<cat-allocation,cat allocation command>>.

Reviewer: I'd suggest including this API call in the docs, similar to the desired balance API ones, so that users can just click on it.

Also could we make any simple additions to `GET _cat/allocation` to make it more useful in this context? For instance, would number of shards on undesired nodes, broken down by node, be a useful thing to include?

Author: Possibly, I was attempting a similar change before, but it is a little verbose and require node stats changes as well, see https://github.com/elastic/elasticsearch/pull/97561/files#r1259473590

Reviewer: I don't think we'd need to do that for the number-of-shards-to-be-moved stat, we could either just execute the `TransportGetDesiredBalanceAction` or else introduce another master-node action dedicated to this task.

Author: Let me see if I can add that in https://github.com/elastic/elasticsearch/pull/105894

TransportGetDesiredBalanceAction is possible, but I would prefer to have cat/node stats rather then internal api

Reviewer: Sure, but cat nodes and cat allocation already call several transport actions, I meant we could add `TransportGetDesiredBalanceAction` under the hood there

[[troubleshooting-unbalanced-cluster]]
== Troubleshooting an unbalanced cluster

> Elasticsearch is balancing shards across data tier to achieve a good compromise between:

Reviewer 3: ```suggestion
Elasticsearch balances shards across data tiers to achieve a good compromise between:
```

* shard count
* disk usage
* ingest load forecast for write indices

> Elasticsearch does not take into account amount or complexity of search queries when rebalancing shards.

Reviewer 3: ```suggestion
Elasticsearch does not take into account the amount or complexity of search queries when rebalancing shards.
```

> There is no guarantee that individual components are going to be evenly spread across the nodes.

Reviewer 3: ```suggestion
There is no guarantee that individual components will be evenly spread across the nodes.
```

> This might appear as some of the nodes have less shards allocated or using less disk
> as long as they are assigned with the shards with greater ingest load forecast.

Reviewer 3: ```suggestion
This could happen if some nodes have fewer shards, or are using less disk space, 
but are assigned shards with higher ingest load forecasts.
```

This sentence was quite hard to read, not sure if this edit makes it clearer. If not, consider breaking into smaller sentences.

> Use <<cat-allocation,cat allocation command>> to list workloads per node:

Reviewer 3: ```suggestion
Use the <<cat-allocation,cat allocation command>> to list workloads per node:
```

[source,console]
--------------------------------------------------
GET /_cat/allocation?v
--------------------------------------------------
// TEST[s/^/PUT test\n{"settings": {"number_of_replicas": 0}}\n/]

The API returns the following response:

[source,text]
--------------------------------------------------
shards shards.undesired write_load.forecast disk.indices.forecast disk.indices disk.used disk.avail disk.total disk.percent host      ip        node    node.role
     1                0                 0.0                  260b         260b    47.3gb     43.4gb    100.7gb           46 127.0.0.1 127.0.0.1 CSUXak2 himrst
--------------------------------------------------
// TESTRESPONSE[s/\d+(\.\d+)?[tgmk]?b/\\d+(\\.\\d+)?[tgmk]?b/ s/46/\\d+/]
// TESTRESPONSE[s/CSUXak2 himrst/.+/ non_json]

> This response contains following components that influence the balancing

Reviewer 3: ```suggestion
This response contains the following information that influences balancing:
```

* `shards` is the current number of shards allocated to the node
* `shards.undesired` is the number of shards that needs to be moved to other nodes to finish balancing

> * `disk.indices.forecast` expected disk usage according to the projected shard growth

Reviewer 3: ```suggestion
* `disk.indices.forecast` is the expected disk usage according to projected shard growth
```

> * `write_load.forecast` project ingest load associated with this node

Reviewer 3: ```suggestion
* `write_load.forecast` is the projected ingest load associated with this node
```

nit: consistency

A cluster is considered balanced when no shard movements are scheduled (all `shards.undesired` are equal to 0).

> At this state it should have spread workload reasonably (eg. some nodes might have fewer shards but more ingest load).

Reviewer 3: ```suggestion
At this state it should have spread workload reasonably. For example, some nodes might have fewer shards but more ingest load.
```

> Some operations such as node restarting or decommissioning or changing cluster allocation settings

Reviewer 3: ```suggestion
Some operations such as node restarting, decommissioning, or changing cluster allocation settings
```

are disruptive and might require multiple shards to move or rebalance.

Shard movement order is not deterministic and mostly determined by the source and target node readiness to move a shard.
While rebalancing is in progress some nodes might appear busier then others.

> When shard is allocated to undesired node it utilizes the resources of the current node instead of target.

Reviewer 3: ```suggestion
When a shard is allocated to an undesired node it uses the resources of the current node instead of target.
```

> This might cause a hotspot (disk or CPU) when multiple shards are residing on the current node that have not been

Reviewer 3: ```suggestion
This might cause a hotspot (disk or CPU) when multiple shards reside on the current node that have not been
```

moved to their corresponding targets yet.

> If cluster is taking a long time to finish rebalancing you might find following log entries:

Reviewer 3: ```suggestion
If cluster takes a long time to finish rebalancing you might find the following log entries:
```

[source,text]
--------------------------------------------------
[WARN][o.e.c.r.a.a.DesiredBalanceReconciler] [10%] of assigned shards (10/100) are not on their desired nodes, which exceeds the warn threshold of [10%]
--------------------------------------------------
This is not concerning as long as the number of such shards is decreasing and this warning appears occasionally,
for example after rolling restarts or changing allocation settings.

> If cluster is experiencing this warning for extended period of time, it is possible that the desired balance diverged

Reviewer 3: ```suggestion
If the cluster has this warning for an extended period of time, it is possible that the desired balance has diverged
```

Reviewer 3: Maybe we could be more quantitative than "an extended period of time" here?

Author: I would not mind, but I am not sure how to quantify it as it depends on many things such as cluster size, amount/size of shards, background load, etc

> too far from the current state. In such case is possible to reset the desired balance using the following API call
> and calculate a new one that would hopefully require less shard movements to get to balanced state.

Reviewer 3: ```suggestion
too far from the current state. In such cases you can reset the desired balance and calculate a 
new one that requires less shard movements to get to balanced state. Reset the desired balance 
using the following API call:
```

[source,console,id=delete-desired-balance-request-example]
--------------------------------------------------
DELETE /_internal/desired_balance
--------------------------------------------------

> If cluster is getting into this state repeatedly without node restarts nor changes in allocation settings,
> it is possible to increase the <<shards-rebalancing-heuristics,cluster.routing.allocation.balance.threshold>> in order
> to reduce the sensitivity of the algorithm that tries to level up the shard count and disk usage within the cluster.

Reviewer 3: ```suggestion
If the cluster gets into this state repeatedly, without node restarts or changes in allocation settings,
you can increase the <<shards-rebalancing-heuristics,`cluster.routing.allocation.balance.threshold`>>.
This will reduce the sensitivity of the algorithm that tries to level up the shard count and disk usage within the cluster.
```

[[troubleshooting-unbalanced-cluster]]
== Troubleshooting an unbalanced cluster

Elasticsearch balances shards across data tiers to achieve a good compromise between:

* shard count
* disk usage

> * ingest load forecast for write indices

Reviewer: ```suggestion
* write load (for indices in data streams)
```

> Elasticsearch does not take into account the amount or complexity of search queries when rebalancing shards.

Reviewer: Maybe mention that shard count is a proxy for this factor?

There is no guarantee that individual components will be evenly spread across the nodes.
This could happen if some nodes have fewer shards, or are using less disk space, 

> but are assigned shards with higher ingest load forecasts.

Reviewer: ```suggestion
but are assigned shards with higher write loads.
```

[source,console]
--------------------------------------------------
GET /_cat/allocation?v
--------------------------------------------------
// TEST[s/^/PUT test\n{"settings": {"number_of_replicas": 0}}\n/]

The API returns the following response:

[source,text]
--------------------------------------------------
shards shards.undesired write_load.forecast disk.indices.forecast disk.indices disk.used disk.avail disk.total disk.percent host      ip        node    node.role
     1                0                 0.0                  260b         260b    47.3gb     43.4gb    100.7gb           46 127.0.0.1 127.0.0.1 CSUXak2 himrst
--------------------------------------------------
// TESTRESPONSE[s/\d+(\.\d+)?[tgmk]?b/\\d+(\\.\\d+)?[tgmk]?b/ s/46/\\d+/]
// TESTRESPONSE[s/CSUXak2 himrst/.+/ non_json]

This response contains the following information that influences balancing:

* `shards` is the current number of shards allocated to the node
* `shards.undesired` is the number of shards that needs to be moved to other nodes to finish balancing
* `disk.indices.forecast` is the expected disk usage according to projected shard growth

> * `write_load.forecast` is the projected ingest load associated with this node

Reviewer: ```suggestion
* `write_load.forecast` is the projected total write load associated with this node
```

> A cluster is considered balanced when no shard movements are scheduled (all `shards.undesired` are equal to 0).

Reviewer: ```suggestion
A cluster is considered balanced when all shards are in their desired locations, which means that no further shard movements are planned (all `shards.undesired` values are equal to 0).
```

> At this state it should have spread workload reasonably. For example, some nodes might have fewer shards but more ingest load.

Reviewer: I'd be inclined to remove this sentence.

Some operations such as node restarting, decommissioning, or changing cluster allocation settings

> are disruptive and might require multiple shards to move or rebalance.

Reviewer: ```suggestion
are disruptive and might require multiple shards to move in order to rebalance the cluster.
```

Shard movement order is not deterministic and mostly determined by the source and target node readiness to move a shard.
While rebalancing is in progress some nodes might appear busier then others.

> When a shard is allocated to an undesired node it uses the resources of the current node instead of target.

Reviewer: ```suggestion
When a shard is allocated to an undesired node it uses the resources of the current node instead of the target.
```

This might cause a hotspot (disk or CPU) when multiple shards reside on the current node that have not been
moved to their corresponding targets yet.

If a cluster takes a long time to finish rebalancing you might find the following log entries:
[source,text]
--------------------------------------------------
[WARN][o.e.c.r.a.a.DesiredBalanceReconciler] [10%] of assigned shards (10/100) are not on their desired nodes, which exceeds the warn threshold of [10%]
--------------------------------------------------
This is not concerning as long as the number of such shards is decreasing and this warning appears occasionally,
for example after rolling restarts or changing allocation settings.

If the cluster has this warning for an extended period of time, it is possible that the desired balance has diverged

> too far from the current state. In such cases you can reset the desired balance and calculate a 

Reviewer: I'd drop the "you can" here, make it an instruction:

```suggestion
too far from the current state. If so, reset the desired balance to calculate a 
```

> new one that requires less shard movements to get to balanced state. Reset the desired balance 

Reviewer: ```suggestion
new one that requires fewer shard movements to get to balanced state. Reset the desired balance 
```

using the following API call:

[source,console,id=delete-desired-balance-request-example]
--------------------------------------------------
DELETE /_internal/desired_balance
--------------------------------------------------

> If the cluster gets into this state repeatedly, without node restarts or changes in allocation settings,
> you can increase the <<shards-rebalancing-heuristics,`cluster.routing.allocation.balance.threshold`>>.

Reviewer: Likewise here, no need for "you can".

```suggestion
If the cluster gets into this state repeatedly, without node restarts or changes in allocation settings,
increase the <<shards-rebalancing-heuristics,`cluster.routing.allocation.balance.threshold`>>.
```

Moreover I'd be tempted to suggest this first.
